Issue179

Kelnor277 opened the pull request:
Submitting PR just to show the changes. There's still problems and I think it can be cleaned up a bit by someone that knows more about CCL.
Fixes Social Relax Jobs with prohibited drugs.
Also fixes Social Relax jobs with Automated Factories.

Automated Factories still do not work with meals.
Looks like the issues is JobGiver_GetFood puts the factory object in foodDef instead of the meal. This breaks the FoodUtility.WillIngestStackCountOf call with a null reference error..

Kelnor277 committed:
Working on Issue #179. Fixed issue where pawns would social relax with any drug.
Moved the Thing predicate in TryGiveJobInt to TryFindIngestibleToNurse.
Additional work required to use automated factories. I'm not sure how to implement that. Left the factory checks in the predicate for now.

Kelnor277 committed:
Adjusted automatic factories to not dispense drugs not on pawn's drug policy. Adjusted social relax to only allow drugs on the drug policy

Kelnor277 committed:
Working on Issue #179. Fixed issue where pawns would social relax with any drug.
Updated AutomatedFactory to have a Dictionary that stores the Pawn's requested thingDef.
Updated JobDriver_SocialRelax's will finalize Injest now as long as a building or injestible is part of the job. (was not injesting synthesized items.)
Updated JoyGiver_SocialRelax  maxNumToCarry to be the desired amount to injest if pawn is using a building to get drugs. TryFindIngestibleToNurse now looks for buildings or stacks of drugs.

Kelnor277 committed:
Updating my test Automatic Factory. Used to validate the code.

ForsakenShell commented on DLL_Project/Toils/Toils_FoodSynthesizer.cs:
```diff
@@ -58,7 +58,37 @@ private static Toil TakeFromSynthesier( TargetIndex ind, Pawn eater, Func<ThingD
+            CCL_Log.Message("Take from drugs.");
+            //would like to put this in the FoodSynthesis class but this would require changing the allowed function
+            //signature to take pawn as well. And a full refactor of all of the functions.
+            Func<ThingDef, bool> validator = FoodSynthesis.IsDrug;
```
If a full refactor is required to make it work then it may be worth it.  Currently no public mods use this functionality anyway so no mods will be losing anything.

ForsakenShell commented on DLL_Project/Toils/Toils_FoodSynthesizer.cs:
```diff
@@ -58,7 +58,37 @@ private static Toil TakeFromSynthesier( TargetIndex ind, Pawn eater, Func<ThingD
-            return TakeFromSynthesier( ind, eater, FoodSynthesis.IsDrug, FoodSynthesis.SortDrug );
+            CCL_Log.Message("Take from drugs.");
+            //would like to put this in the FoodSynthesis class but this would require changing the allowed function
+            //signature to take pawn as well. And a full refactor of all of the functions.
```
If a full refactor is required to make it work then it may be worth it. Currently no public mods use this functionality anyway so no mods will be losing anything.

ForsakenShell commented on _Mod/Modders Resource/YayoFactory/Defs/ThingDefs/SynthYayo.xml:
```diff
@@ -0,0 +1,201 @@
+<?xml version="1.0" encoding="utf-8" ?>
+<SynthYayo>
+
+    <ThingDef ParentName="MakeableDrugBase">
```
"MakeableDrugBase" has not been defined in this mod.  Be sure to define all abstracts you use, they are scoped to the mod that defines them.

ForsakenShell commented on _Mod/Modders Resource/YayoFactory/About/About.xml:
```diff
@@ -0,0 +1,15 @@
+
+    <description>v0.15.0
+
+Test only
```
Don't understand, what is this testing?  The XML with all the Defs is unclear as to the purpose of this mod.

ForsakenShell commented on DLL_Project/Detours/JobDriver_SocialRelax.cs:
```diff
@@ -67,11 +67,13 @@ internal bool HasIngestible
             get
             {
-                var thing = this.TargetThing( OptionalIngestibleInd );
+                var thing = this.CurJob.GetTarget(TargetIndex.C).Thing;
```
Please use named constants, that's what they are there for.

Kelnor277 replied:
No problem.

ForsakenShell commented on DLL_Project/Detours/JobDriver_SocialRelax.cs:
```diff
@@ -67,11 +67,13 @@ internal bool HasIngestible
+                var thing = this.CurJob.GetTarget(TargetIndex.C).Thing;
                 if( thing == null )
                 {
+                    CCL_Log.Message("Thing is null");
```
See other source files for formatting messages for errors using the Trace() method.

Kelnor277 replied:
That's a debug message I missed during clean up. I'll remove it.

Kelnor277 commented:
I think I just had a better idea. Make a SynthesizedThing object extending thing with two additional fields. ThingDefToProduce, and Synthesizer.
The overridden thing functions return the Factory's properties or the ThingDefToProduce ones in ways that make sense. IE. location functions return stuff to path to the factory interaction cell, and def returns the def of the production item.

That way we just return this object and no need to store the thingdef in a dictionary. Might even work without modifying base code, or at least require fewer modifications.

Kelnor277 committed:
Fixed null ref exception by populating foodDef with the factories BestProduct call.

Kelnor277 committed:
Added code to remove the dictionary entry created by the joygiver so the pawn.
Found ThingDef Job.plantsToSow that I can use to store the drug def.
New class FactoryWithProducts extending Thing. Holds the drug def (And additional potential values for the future.) long enough to pass back to the tryGiveJobInt function.
Changed the start carry logic to use the passed TargetIndex instead of Target.C. That broke the get food job since the ingestible is Target.A there.
Food does work here. But I think the logic might have some bugs in it and we should detour the FoodUtility to control the find best food source logic.

ForsakenShell commented on DLL_Project/Detours/JobGiver_GetFood.cs:
```diff
@@ -71,13 +71,21 @@ protected override Job TryGiveJob( Pawn pawn )
+                    //Not an issue with nutrient paste, since if you can't fill a nutrient paste hopper you're probably starving anyway.
+                    //But if you have a factory that makes Lavish Meals from meat. Colonists will wander around starving
+                    //trying to get meat even if there's fine meals in storage. I think anyway.
+                    //Also, current code would ignore a Fine meal factory if the Lavish meal factory has unreachable hoppers.
```
BestProduct() looks at the entire list of available recipes and all the resources available.  If it doesn't return a null then the def it returns can indeed be produced and calling HasEnoughResourceInHoppersFor() for the def returned is redundant.  If BestProduct() returns null then the factory either has no recipes or not enough resources available for any recipes it does have.

If BestProduct() returns a def when there isn't the resources to produce it then BestProduct() has a bug which needs to be fixed.

Kelnor277 replied:
For now I'm just using BestProduct as a band-aid to populate foodDef to prevent a null exception. I'll double check, but I was getting null exceptions and I'm pretty sure I traced it to foodDef coming back from TryFindBestFoodSourceFor as the factories' def, not the food itself. It looks like TryFindBestFoodSourceFor needs to be detoured to properly fix that.
As for the comment, I see your point and it's not an issue using BestProduct.
Edit. The other possibility is to figure out how TryFindBestFoodSourceFor get's the foodDef and make the factory play along with it without detouring. Unfortunately my decompiler isn't playing nice with that method.

ForsakenShell replied:
TryFindBestFoodSourceFor() works by calling several methods depending on the situation.  The situation we care about is when it calls BestFoodSourceOnMap() which has been detoured to handle factories.  Best have your IDE find all references to Building_AutomatedFactory.  That will show you all the detours, etc you need which are already in place but will need updating for this and every alpha since they were first implemented (back in ~A12 when factories were added).  To handle this overlap in systems (drugs and food), we may need to step back and refactor how factories are coded to play better.

Kelnor277 committed:
Uncommented the automated factory SocialRelax code, issue is fixed.
Left other drug related code commented. Those issues haven't been resolved.

Kelnor277 committed:
Removed the comment about the hopper filling.
